custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go

```diff
@@ -0,0 +1,163 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	stackdriver "google.golang.org/api/monitoring/v3"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/pkg/api"
+
+	// Install registers the API group and adds types to a scheme.
+	_ "k8s.io/client-go/pkg/api/install"
+	"k8s.io/metrics/pkg/apis/custom_metrics"
+
+	"github.com/GoogleCloudPlatform/k8s-stackdriver/custom-metrics-stackdriver-adapter/pkg/config"
+)
+
+// Translator is a structure used to translate between Custom Metrics API and Stackdriver API
+type Translator struct {
+	service   *stackdriver.Service
+	config    *config.GceConfig
+	reqWindow time.Duration
+	clock     clock
+}
+
+// GetSDReqForNamespacedObject returns Stackdriver request for query for a single namespaced object.
+func (t *Translator) GetSDReqForNamespacedObject(object runtime.Object, metricName string) (*stackdriver.ProjectsTimeSeriesListCall, error) {
+	objMeta := object.(metav1.ObjectMetaAccessor).GetObjectMeta()
+	resourceID := fmt.Sprintf("%s", objMeta.GetUID())
+
+	if objMeta.GetClusterName() != t.config.Cluster {
+		return nil, fmt.Errorf("Cluster name mismatch - found pod with cluster name %s, but cluster name %s received from Metadata Server.", objMeta.GetClusterName(), t.config.Cluster)
+	}
+	project := fmt.Sprintf("projects/%s", t.config.Project)
+	endTime := t.clock.Now()
+	startTime := endTime.Add(-t.reqWindow)
+	filter, err := t.createFilter([]string{fmt.Sprintf("\"%s\"", resourceID)}, t.config.Project, t.config.Zone, t.config.Cluster, metricName)
+	if err != nil {
+		return nil, err
+	}
+	return t.service.Projects.TimeSeries.List(project).Filter(filter).
+		IntervalStartTime(startTime.Format("2006-01-02T15:04:05Z")).
+		IntervalEndTime(endTime.Format("2006-01-02T15:04:05Z")).
+		AggregationPerSeriesAligner("ALIGN_NEXT_OLDER").
+		AggregationAlignmentPeriod(fmt.Sprintf("%vs", int64(t.reqWindow.Seconds()))), nil
+}
+
+// GetRespForNamespacedObject returns translates Stackdriver response to a Custom Metric associated
+// with a single namespaced object.
+func (t *Translator) GetRespForNamespacedObject(response *stackdriver.ListTimeSeriesResponse, groupResource schema.GroupResource, metricName string, namespace string, name string) (*custom_metrics.MetricValue, error) {
+	values, err := t.getMetricValuesFromResponse(groupResource, namespace, response)
+	if err != nil {
+		return nil, err
+	}
+	if len(values) != 1 {
+		return nil, fmt.Errorf("Expected exactly one value for pod %s in namespace %s, but received %v values", name, namespace, len(values))
+	}
+	for _, value := range values {
```

Comment:
I think this way it's a little bit cleaner than if I added pod id as another argument to the function. Added a comment to make this a little bit more readable.